azure-iot-device/tests/iothub/transport/mqtt/test_mqtt_provider.py

```diff
@@ -8,164 +8,1120 @@
-def test_disconnect_calls_loopstop_on_mqttclient(MockMqttClient):
-    mock_mqtt_client = MockMqttClient.return_value
-
-    mqtt_provider = MQTTProvider(fake_device_id, fake_hostname, fake_username)
-    mqtt_provider.disconnect()
-
-    mock_mqtt_client.disconnect.assert_called_once_with()
-
-
-@patch.object(mqtt, "Client")
-def test_publish_calls_publish_on_mqtt_client(MockMqttClient):
-    mock_mqtt_client = MockMqttClient.return_value
-    mock_mqtt_client.publish = MagicMock(return_value=mqtt.MQTTMessageInfo(fake_mid))
-
-    topic = "topic/"
-    event = "Tarantallegra"
-
-    mqtt_provider = MQTTProvider(fake_device_id, fake_hostname, fake_username)
-    pub_mid = mqtt_provider.publish(topic, event)
-
-    assert pub_mid == fake_mid
-    mock_mqtt_client.publish.assert_called_once_with(topic=topic, payload=event, qos=1)
-
-
-@patch.object(mqtt, "Client")
-def test_reconnect_calls_username_pw_set_and_reconnect_on_mqtt_client(MockMqttClient):
-    mock_mqtt_client = MockMqttClient.return_value
-
-    mqtt_provider = MQTTProvider(fake_device_id, fake_hostname, fake_username)
-    mqtt_provider.reconnect(new_fake_password)
-
-    mock_mqtt_client.username_pw_set.assert_called_once_with(
-        username=fake_username, password=new_fake_password
-    )
-    mock_mqtt_client.reconnect.assert_called_once_with()
-
-
-@patch.object(mqtt, "Client")
-def test_subscribe_calls_subscribe_on_mqtt_client(MockMqttClient):
-    mock_mqtt_client = MockMqttClient.return_value
-    mock_mqtt_client.subscribe = MagicMock(return_value=(fake_rc, fake_mid))
-
-    mqtt_provider = MQTTProvider(fake_device_id, fake_hostname, fake_username)
-    sub_mid = mqtt_provider.subscribe(fake_topic, fake_qos)
-
-    assert sub_mid == fake_mid
-    mock_mqtt_client.subscribe.assert_called_once_with(fake_topic, fake_qos)
-
-
-@patch.object(mqtt, "Client")
-def test_unsubscribe_calls_unsubscribe_on_mqtt_client(MockMqttClient):
-    mock_mqtt_client = MockMqttClient.return_value
-    mock_mqtt_client.unsubscribe = MagicMock(return_value=(fake_rc, fake_mid))
-
-    mqtt_provider = MQTTProvider(fake_device_id, fake_hostname, fake_username)
-    unsub_mid = mqtt_provider.unsubscribe(fake_topic)
+class DummyException(Exception):
+    pass
+
+
+@pytest.fixture
+def mock_mqtt_client(mocker):
+    mock = mocker.patch.object(mqtt, "Client")
+    mock_mqtt_client = mock.return_value
+    mock_mqtt_client.subscribe = mocker.MagicMock(return_value=(fake_rc, fake_mid))
+    mock_mqtt_client.unsubscribe = mocker.MagicMock(return_value=(fake_rc, fake_mid))
+    return mock_mqtt_client
+
+
+@pytest.fixture
+def provider(mock_mqtt_client):
+    # Implicitly imports the mocked Paho MQTT Client from mock_mqtt_client
+    return MQTTProvider(client_id=fake_device_id, hostname=fake_hostname, username=fake_username)
+
+
+@pytest.mark.describe("MQTT Provider - Instantiation")
+class TestInstantiation(object):
+    @pytest.mark.it("Creates an instance of the Paho MQTT Client")
```

Comment:
>@pytest.mark.it("Creates an instance of the Paho MQTT Client") [](start = 4, length = 62)

Awesome!  I love this.  using the pytest.mark.it decorator is so much better than having the test name generated from the name of the test function.

Reply:
Shout out to the two guys maintaining pytest-testdox for pushing through features and fixes when I ask.

---
In reply to: [277844525](https://github.com/Azure/azure-iot-sdk-python-preview/pull/82#discussion_r277844525) [](ancestors = 277844525)